kornia/filters/filter.py

```diff
@@ -71,16 +91,119 @@ def filter2D(input: torch.Tensor, kernel: torch.Tensor,
 
     # prepare kernel
     b, c, h, w = input.shape
-    tmp_kernel: torch.Tensor = kernel.unsqueeze(0).to(input.device).to(input.dtype)
+    tmp_kernel: torch.Tensor = kernel[None].type_as(input)
+
     if normalized:
         tmp_kernel = normalize_kernel2d(tmp_kernel)
+
     # pad the input tensor
     height, width = tmp_kernel.shape[-2:]
-    padding_shape: List[int] = compute_padding((height, width))
+    padding_shape: List[int] = compute_padding([height, width])
     input_pad: torch.Tensor = F.pad(input, padding_shape, mode=border_type)
-    b, c, hp, wp = input_pad.shape
+
     # convolve the tensor with the kernel. Pick the fastest alg
     kernel_numel: int = height * width
     if kernel_numel > 81:
+        b, c, hp, wp = input_pad.shape
         return F.conv2d(input_pad.reshape(b * c, 1, hp, wp), tmp_kernel, padding=0, stride=1).view(b, c, h, w)
+
     return F.conv2d(input_pad, tmp_kernel.expand(c, -1, -1, -1), groups=c, padding=0, stride=1)
```

Comment:
Should we allow direct input of channel-wise-kernel as well?

Reply:
not sure what do you mean ? /cc @ducha-aiki knows more about this function :)

Reply:
Well, if someone has channel-wise kernel, i.e. not a single kernel for all channels, but valid kernel for conv2d, then there is no point of using kornia.filter2d. One should use directly F.conv2d then